dataplane/server.go

```diff
@@ -24,3 +35,57 @@ const (
 type Server struct {
 	dpb.UnimplementedHALServer
 }
+
+var (
+	// Stubs for testing
+	setInterfaceHWAddr = kernel.SetInterfaceHWAddr
+	setInterfaceIPs    = kernel.SetInterfaceIPs
+	setInterfaceState  = kernel.SetInterfaceState
+)
+
+// UpdatePort updates an interface properties from the input request.
+func (s *Server) UpdatePort(_ context.Context, req *dpb.UpdatePortRequest) (*dpb.UpdatePortResponse, error) {
+	if req.Name == "" {
+		return nil, status.Error(codes.InvalidArgument, "name not set")
+	}
+	tapName := fmt.Sprintf("%s-tap", req.Name)
+	if req.Hwaddr != "" {
+		addr, err := net.ParseMAC(req.Hwaddr)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "invalid MAC address %q: parse err %v", req.Hwaddr, err)
+		}
+		if err := setInterfaceHWAddr(tapName, addr); err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to set HW addr: %v", err)
+		}
+	}
+	var ips []*net.IPNet
+	for _, ip := range req.Ipv4S {
+		parsedIP := net.ParseIP(ip.GetIp())
+		if parsedIP == nil {
+			return nil, status.Errorf(codes.InvalidArgument, "invalid IP address: %q", ip.GetIp())
+		}
+		ips = append(ips, &net.IPNet{
+			IP:   parsedIP,
+			Mask: net.CIDRMask(int(ip.GetPrefixLen()), 32),
+		})
+	}
+	for _, ip := range req.Ipv6S {
+		parsedIP := net.ParseIP(ip.GetIp())
+		if parsedIP == nil {
+			return nil, status.Errorf(codes.InvalidArgument, "invalid IP address: %q", ip.GetIp())
+		}
+		ips = append(ips, &net.IPNet{
+			IP:   parsedIP,
+			Mask: net.CIDRMask(int(ip.GetPrefixLen()), 128),
+		})
+	}
+	if err := setInterfaceIPs(tapName, ips); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to set IPs: %v", err)
+	}
+	if req.AdminState != dpb.PortState_PORT_STATE_UNSPECIFIED {
+		if err := setInterfaceState(tapName, req.AdminState == dpb.PortState_PORT_STATE_UP); err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to set interface set: %v", err)
+		}
+	}
```

Comment:
switch statement in case we add another enum value?

Reply:
done